pbo/src/discrete.py

```diff
@@ -0,0 +1,154 @@
+# Custom imports
+from pbo.src.base import *
+
+###############################################
+### Class discrete
+### A PBO agent for discrete optimization
+class discrete(base_pbo):
+    def __init__(self, params, act_dim, obs_dim):
+
+        # Initialize from arguments
+        self.pdf        = params.pdf
+        self.act_dim    = len(act_dim) # number of parameters
+        self.obs_dim    = obs_dim
+        self.param_dim  = act_dim # list containing the dimension for each param
+        self.n_gen      = params.n_gen
+        self.n_ind      = params.n_ind
+        self.size       = self.n_gen*self.n_ind
+        self.n_cpu      = params.n_cpu
+
+        self.param_gen        = params.param_gen
+        self.batch            = params.batch
+        self.lr               = params.lr
+        self.epochs           = params.epochs
+        self.adv_clip         = params.adv_clip
+        self.adv_decay        = params.adv_decay
+        self.continuity_regul = params.continuity_regul
+
+        # Build network
+        self.net = nn(params.arch,
+                      sum(self.param_dim),
+                      'relu',
+                      None,
+                      self.lr)
+
+        # Init network parameters
+        dummy = self.net(tf.ones([1,self.obs_dim]))
+
+        # Storing buffers
+        self.idx     = 0
+
+        self.gen     = np.zeros( self.size,               dtype=np.int32)
+        self.ep      = np.zeros( self.size,               dtype=np.int32)
+        self.obs     = np.zeros((self.size, self.obs_dim),dtype=np.float64)
+        self.act     = np.zeros((self.size, self.act_dim),dtype=np.float64)
+        self.acc     = np.zeros((self.size, self.act_dim),dtype=np.float64)
+        self.adv     = np.zeros( self.size,               dtype=np.float64)
+        self.rwd     = np.zeros( self.size,               dtype=np.float64)
+
+        self.bst_gen = np.zeros( self.n_gen,              dtype=np.int32)
+        self.bst_ep  = np.zeros( self.n_gen,              dtype=np.int32)
+        self.bst_acc = np.zeros((self.n_gen,self.act_dim),dtype=np.float64)
+        self.bst_rwd = np.zeros( self.n_gen,              dtype=np.float64)
+        self.avg_rwd = np.zeros( self.n_gen,              dtype=np.float64)
+
+    # Get actions from network
+    def get_actions(self, state, n):
+
+        # Predict
+        x  = tf.convert_to_tensor([state[0]], dtype=tf.float64)
+        cat_params = self.net.call(x)
+        cat_params = np.asarray(cat_params)[0] # take first worker only for extracting the params
+
+        # Build pdfs
+        pdfs  = [None]*self.act_dim
+        index = 0
+        for i in range(self.act_dim):
+            pdfs[i] = tfd.Categorical(cat_params[index:index + self.param_dim[i]])
+            index  += self.param_dim[i]
+
```

Comment:
@aurelegoetz it seems to me that self.param_dim is not a list ?

Reply:
It is, check in sphere_2d_discrete.py at line 18: 
self.act_size = [len(val) for val in self.x_values]
which is then passed to the constructor of the discrete object.